### PR TITLE
Coop proto5

### PR DESCRIPTION
**Pull request by mhowlett**

### CODE REVIEW

File: src-cpp/rdkafkacpp.h [hunk 1]
```diff
@@ -2668,6 +2668,23 @@ class RD_EXPORT KafkaConsumer : public virtual Handle {
...
+   *        online migration of a consumer group from using one
+   *        protocol to another (in particular upgading from EAGER
+   *        to COOPERATIVE) without a restart is not currently
+   *        supported.
+   *
+   * @returns an empty string on error, or one of
+   *          "NONE", "EAGER", "COOPERATIVE" on success.
+   */
+
+  virtual std::string rebalance_protocol () = 0;
```
Comment (mhowlett) [inline src-cpp/rdkafkacpp.h:2686]: an enum is the more appropriate thing to be using, but it's not like that's really much more type safe in C / C++ than a string, particularly with interop code, so not much better, so yeah string, whatever.

File: src/rd.h [hunk 2]
```diff
@@ -100,6 +100,15 @@
 /** Assert if reached */
 #define RD_NOTREACHED() rd_assert(!*"/* NOTREACHED */ violated")
 
+/** Assert if reached */
+#define RD_BUG(...) do {                                                \
```
Comment (mhowlett) [inline src/rd.h:104]: nice! with that said, I've been wondering about the tradeoff between fail fast resulting in higher quality software and user perceptions (!!!! BUG is probably a worse perception than some subtly undefined behavior which may go unnoticed). i was always strongly on the side of fail fast, but i've started to wonder about it from a broader perspective.

File: src/rdkafka_assignment.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+#define _RDKAFKA_ASSIGNMENT_H_
+
+typedef struct rd_kafka_assignment_s {
+        /** All currently assigned partitions. */
+        rd_kafka_topic_partition_list_t *all;
+        /** Partitions in need of action (subset of .all) */
+        rd_kafka_topic_partition_list_t *pending;
+        /** Partitions that are being queried for committed
+         * offsets (subset of .all) */
+        rd_kafka_topic_partition_list_t *queried;
```
Comment (mhowlett) [inline src/rdkafka_assignment.h:40]: it would be safer if `pending` and `queried` were specified as a function of `all`, or `all` was specified as a function of `pending` and `queried`. In a functional language, writing that is very terse, in C, it seems probably easier to read as you have it, and easier to read is probably the more important thing to optimize on.

so no change, just pushing the benefits of other languages at any opportunity.

File: src/rdkafka_assignment.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        /** Partitions that have been removed from the assignment
+         * but not yet decommissioned. (not included in .all) */
+        rd_kafka_topic_partition_list_t *removed;
+        /** Number of started partitions */
+        int started_cnt;
```
Comment (mhowlett) [inline src/rdkafka_assignment.h:45]: again smells like these should be derived values.

Comment (mhowlett) [reply src/rdkafka_assignment.h:45]: but not proposing changes at this point.

File: src/rdkafka_assignment.c [hunk 5]
```diff
@@ -0,0 +1,939 @@
...
+ * It is important to call rd_kafka_assignment_serve() after each change
+ * to the assignment through assignment_add, assignment_subtract or
+ * assignment_clear as those functions only modify the assignment but does
+ * not take any action to transition partitions to or from the assignment
+ * states.
+ *
+ * The reason assignment_serve() is not automatically called from these
+ * functions is for the caller to be able to set the current state before
+ * the side-effects of serve() kick in, such as the call to
+ * rd_kafka_cgrp_assignment_done() that in turn will set the cgrp state.
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:44]: i don't deeply get this yet, and I expect it's a step closer to having something comprehensible, and about as good as we can do without making bigger changes, but it all smells of complexity still.

File: src/rdkafka_assignment.c [hunk 6]
```diff
@@ -0,0 +1,939 @@
...
+ *
+ * @remark Try to keep any cgrp state out of this file.
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:46]: bad smell...

File: src/rdkafka_assignment.c [hunk 7]
```diff
@@ -0,0 +1,939 @@
...
+ * Called from the FetchOffsets response handler below.
+ */
+void
+rd_kafka_assignment_apply_offsets (rd_kafka_cgrp_t *rkcg,
+                                   rd_kafka_topic_partition_list_t *offsets,
+                                   rd_kafka_resp_err_t err) {
+        rd_kafka_topic_partition_t *rktpar;
+
+        /* Request-level error */
+        if (err)
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:100]: this err is handled in a similar way by the caller as well.
better not to pass in the err as an argument.

File: src/rdkafka_partition.h [hunk 8]
```diff
@@ -561,6 +580,36 @@ rd_kafka_topic_partition_list_upsert (
...
+void rd_kafka_topic_partition_list_add_list (
+        rd_kafka_topic_partition_list_t *dst,
+        const rd_kafka_topic_partition_list_t *src);
+
+/**
+ * Traverse rd_kafka_topic_partition_list_t.
+ *
+ * @warning \p TPLIST modifications are not allowed.
+ */
+#define RD_KAFKA_TPLIST_FOREACH(RKTPAR,TPLIST)                          \
```
Comment (mhowlett) [inline src/rdkafka_partition.h:597]: nice

File: src/rdkafka_assignment.h [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+#ifndef _RDKAFKA_ASSIGNMENT_H_
+#define _RDKAFKA_ASSIGNMENT_H_
+
+typedef struct rd_kafka_assignment_s {
```
Comment (mhowlett) [inline src/rdkafka_assignment.h:33]: it would be helpful to have really verbose description of what each of the lists are / when they get updated / anything. i'm finding i want to naturally come back here to solidify that.

File: src/rdkafka_assignment.c [hunk 10]
```diff
@@ -0,0 +1,939 @@
...
+        /* Apply the fetched offsets to the assignment */
+        rd_kafka_assignment_apply_offsets(rkcg, offsets, err);
+
+        rd_kafka_topic_partition_list_destroy(offsets);
+}
+
+
+/**
+ *
+ * - Hold off starting partitions if there are outstanding commits.
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:233]: incorporate in function docstring

File: src/rdkafka_assignment.c [hunk 11]
```diff
@@ -0,0 +1,939 @@
...
+}
+
+/**
+ * @brief Apply the fetched committed offsets to the current assignment's
+ *        queried partitions.
+ *
+ * Called from the FetchOffsets response handler below.
+ */
+void
+rd_kafka_assignment_apply_offsets (rd_kafka_cgrp_t *rkcg,
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:94]: this function signature seems innocent, but it's very complex because you pass in `rkcg`, which means you can do anything in the function you like (and not just to the cg, there is an `rkcg_rk` as well. rkcg shouldn't have this). as a reader of the code, i have to keep the possibility that the function does anything open in my head. i often i have a fuzzy idea of how things work when i'm trying to read through and understand it, so even if the meaning is clear to you, it is not clear enough to me that i don't have to check. if instead this signature looked something like this:

```
void
rd_kafka_assignment_apply_offsets (rd_kafka_assignment_t *assignment, rd_kafka_topic_partition_list_t *offsets, rd_kafka_q_t *app_poll_queue, logger_t *logger)
```

I have quite a lot of certainty over what the function can do just from the function signature. i can make assumptions and they are probably going to be right.

If all of librdkafka was changed to work this way, complexity drops dramatically.

Comment (edenhill) [reply src/rdkafka_assignment.c:94]: You can think of the recurring `rkcg` argument as `this` or `self` in OO-languages, the assignment resides on the rkcg and while we could have an assignment->rkcg backpointer it wouldn't really serve any purpose in compartmentalization since the assignment and rkcg are somewhat intertwined: I did consider making them completely separate, but that would require an abstraction for inter rkcg-assignment communication with a miniscule ROI.

Comment (mhowlett) [reply src/rdkafka_assignment.c:94]: yes, of course i realize `rkcg` is analogous to `this`. my point is this OO-structure of organizing software is bad. The reason is that this function takes in a lot more state than it needs, and it makes it hard to know for sure what it does from the signature, c.f. what I wrote where it is obvious.

> the assignment resides on the rkcg and while we could have an assignment->rkcg backpointer

no backpointers! dramatic increase in complexity. avoid at all cost.

In the proposed signature, I pass in the other bits of the cgrp needed separately, which although more parameters, makes it easier to understand, because it documents what's used, and you can guess some of the behavior of the fn from what parameters it explicitly takes.

> I did consider making them completely separate

they should be separate in the sense that rkcg should have an assignment_t (which is what you have).

File: src/rdkafka_assignment.c (hunk 5 above)
Comment (edenhill) [reply src/rdkafka_assignment.c:44]: Compare

this
```c
assignment_subtract(p1);
cgrp_set_join_state(wait_incr_unassign_to_complete);
assignment_serve();  // triggers assignmen_done() because p1 was not started and had no offsets to commit
...
assignment_done() {
    if (state == WAIT_INCR_UNASSIGN..)
        transition_to_next_state();
}
``` 
with (what we don't do):
```c
assignment_subtract_that_also_does_serve(p1); // triggers assignmen_done() because p1 was not started and had no offsets to commit
cgrp_set_join_state(wait_incr_unassign_to_complete);
...
assignment_done() {
    if (state == WAIT_INCR_UNASSIGN..) // NOTREACHED since the state hasn't been set yet
        transition_to_next_state();
}
```

File: src/rdkafka_assignment.h (hunk 3 above)
Comment (edenhill) [reply src/rdkafka_assignment.h:40]: librdkafka is still primarily a C project.

File: src/rdkafka_assignment.c [hunk 12]
```diff
@@ -0,0 +1,939 @@
...
+                                RD_KAFKA_OFFSET_INVALID,
+                                "Failed to fetch committed offset for "
+                                "group \"%s\" topic %s [%"PRId32"]: %s",
+                                rkcg->rkcg_group_id->str,
+                                rktpar->topic, rktpar->partition,
+                                rd_kafka_err2str(rktpar->err));
+
+                        /* The partition will not be added back to .pending
+                         * and thus only reside on .all until the application
+                         * unassigns it and possible re-assigns it. */
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:138]: "unassigns it or possibly re-assigns it."

no-retry requirement for this?

Comment (edenhill) [reply src/rdkafka_assignment.c:138]: The retriable errors have already been retried when we get here.

File: src/rdkafka_assignment.c [hunk 13]
```diff
@@ -0,0 +1,939 @@
...
+                                     rktpar->partition,
+                                     rktpar->offset);
+
+                        rd_kafka_topic_partition_list_add_copy(
+                                rkcg->rkcg_assignment.pending, rktpar);
+                }
+        }
+
+        if (offsets->cnt > 0)
+                rd_kafka_assignment_serve(rkcg);
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:167]: would it help the reader if it was always the responsibility of the caller to call `assignment_serve` on these functions?

having thought that, i realized it's related to my comment above in that it makes the function more focused. but where i was coming from was also consistency.

(i'm just brain dumping).

File: src/rdkafka_assignment.c [hunk 14]
```diff
@@ -0,0 +1,939 @@
...
+
+        if (rd_kafka_buf_version_outdated(request, rkcg->rkcg_version)) {
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "OFFSETFETCH",
+                             "Group \"%s\": "
+                             "ignoring outdated OffsetFetch response",
+                             rkcg->rkcg_group_id->str);
+                return;
+        }
+
+        /* If all partitions already had usable offsets then there
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:200]: maybe:

"If all partitions already had usable offsets, then no request will be sent and this handler will be called with `reply` as NULL."

File: src/rdkafka_assignment.c [hunk 15]
```diff
@@ -0,0 +1,939 @@
...
+                                     "(cgrp state %s, awaiting %d commits)",
+                                     rktpar->topic, rktpar->partition,
+                                     rd_kafka_cgrp_state_names[rkcg->
+                                                               rkcg_state],
+                                     rkcg->rkcg_wait_commit_cnt);
+
+                        continue; /* Keep rktpar on pending list */
+                }
+
+                /* Remove rktpar from the pending list */
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:452]: minor: maybe clearer to duplicate this del_by_idx on the two cases that need to it. not at all hard to follow though.

File: src/rdkafka_assignment.c [hunk 16]
```diff
@@ -0,0 +1,939 @@
...
+}
+
+
+/**
+ * @brief Serve all partitions in the pending list.
+ *
+ * This either (asynchronously) queries the partition's committed offset, or
+ * if the start offset is known, starts the partition fetcher.
+ *
+ * @returns >0 if there are pending operations in progress, else 0.
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:352]: maybe be a bit more precise on what the return value means.

File: src/rdkafka_assignment.c [hunk 17]
```diff
@@ -0,0 +1,939 @@
...
+}
+
+
+/**
+ * @returns true if the current or previous assignment has operations in
+ *          progress, such as waiting for partition fetchers to stop.
+ */
+rd_bool_t rd_kafka_assignment_in_progress (rd_kafka_cgrp_t *rkcg) {
+        return rkcg->rkcg_wait_commit_cnt > 0 ||
+                rkcg->rkcg_assignment.wait_stop_cnt > 0 ||
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:548]: the mixing of rkcg and rkcg_assignment is a sign of something not as well factored as it could be i think (not suggesting changes, just brain dump). also elsewhere. what's in this file is all related to assignments, but some of it is operations on assignment, some is more behavior of rkcg.

Comment (edenhill) [reply src/rdkafka_assignment.c:548]: Yeah, I agree but I couldn't find a more suitable solution with the given time constraints.
The problem is that commits may originate from the cgrp/the current assignment, or be completely spontaneous when the application calls commit(some_partitions_that_may_or_may_not_be_part_of_the_assignment).
So this counter had to leak between the abstractions.

File: src/rdkafka_assignment.c [hunk 18]
```diff
@@ -0,0 +1,939 @@
...
+ *
+ * @remark Make sure to call rd_kafka_assignment_serve() after successful
+ *         return from this function.
+ */
+void rd_kafka_assignment_clear (rd_kafka_cgrp_t *rkcg) {
+
+        /* Any change to the assignment marks the current assignment
+         * as not lost.
+         * FIXME: Why is this? */
+        rd_kafka_assignment_clear_lost(rkcg, "assignment removed");
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:566]: well, it is a specific assignment that gets flagged as lost. if the assignment is updated, then we have a new assignment that isn't assumed to be lost anymore (until/unless detected otherwise).

File: src/rdkafka_assignment.c [hunk 19]
```diff
@@ -0,0 +1,939 @@
...
+
+        if (!rkcg->rkcg_assignment.all) {
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:568]: i assume you've gotten rid of this

File: src/rdkafka_assignment.c [hunk 20]
```diff
@@ -0,0 +1,939 @@
...
+                        rktpar->offset = RD_KAFKA_OFFSET_STORED;
+
+                /* Get toppar object for each partition.
+                 * This is to make sure the rktp stays alive while unassigning
+                 * any previous assignment in the call to
+                 * assignment_clear() below. */
+                rd_kafka_topic_partition_ensure_toppar(rkcg->rkcg_rk, rktpar,
+                                                       rd_true);
+
+                /* FIXME: old cgrp_assign() marks rktp as desired, should we? */
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:668]: if so, it belongs outside of this method, which is otherwise pretty much just focused on `rkcg_assignment`

File: src/rdkafka_assignment.c [hunk 21]
```diff
@@ -0,0 +1,939 @@
...
+        rd_kafka_topic_partition_list_clear(rkcg->rkcg_assignment.queried);
+
+        rd_kafka_topic_partition_list_add_list(rkcg->rkcg_assignment.removed,
+                                               rkcg->rkcg_assignment.all);
+        rd_kafka_topic_partition_list_destroy(rkcg->rkcg_assignment.all);
+        rkcg->rkcg_assignment.all = NULL;
+
+        /* Invalidate any outstanding offset fetch requests as a new
+         * assignment with the same partitions may otherwise use the response */
+        rd_kafka_cgrp_version_new_barrier(rkcg);
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:591]: structurally, this mixes in a little bit of CG functionality into a function which is pretty much just `rkcg_assignment`. being ruthless about minimizing state in functions i think pays off in understandability.

File: src/rdkafka_assignment.c [hunk 22]
```diff
@@ -0,0 +1,939 @@
...
+                     rkcg->rkcg_assignment.queried->cnt);
+
+        rd_kafka_wrlock(rkcg->rkcg_rk);
+        rkcg->rkcg_c.assignment_size = rkcg->rkcg_assignment.all->cnt;
+        rd_kafka_wrunlock(rkcg->rkcg_rk);
+
+        /* Any change to the assignment marks the current assignment
+         * as not lost.
+         * FIXME: Why is this? */
+        rd_kafka_assignment_clear_lost(rkcg, "assignment updated");
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:706]: see prev comment on why. 

also same comment that this is a CG related state change in a fn otherwise pretty focused on just rkcg_assignment, good to constrain parameters etc. i think there are a bunch of quick wins throughout librdkafka in doing this. and it's something that can be added very incrementally. i'm sure we can have a good argue about this later.

File: src/rdkafka_assignment.c (hunk 21 above)
Comment (edenhill) [reply src/rdkafka_assignment.c:591]: The barriers have been removed from the next version.

File: src/rdkafka_assignment.c [hunk 23]
```diff
@@ -0,0 +1,939 @@
...
+                if (!rd_kafka_topic_partition_list_find(
+                            rkcg->rkcg_assignment.all,
+                            rktpar->topic, rktpar->partition))
+                        return rd_kafka_error_new(
+                                RD_KAFKA_RESP_ERR__INVALID_ARG,
+                                "%s [%"PRId32"] can't be unassigned since "
+                                "it is not in the current assignment",
+                                rktpar->topic, rktpar->partition);
+
+                rd_kafka_topic_partition_ensure_toppar(rkcg->rkcg_rk, rktpar,
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:756]: more unactionable commentary. I expect I think there shouldn't be a toppar reference in rktpar, rather you'd reference into another collection always. there are other ways in which this method is coupled to CG, generally i would try to factor this so they weren't.

File: src/rdkafka_assignment.c [hunk 24]
```diff
@@ -0,0 +1,939 @@
...
+        if (matched_queried_partitions)
+                rd_kafka_cgrp_version_new_barrier(rkcg);
+
+        /* If the assignment is now empty we remove it.
+         * Empty assignments may only be the result of an assign()
+         * or incremental_assign(), not unassign() or incremental_unassign(). */
+        if (rkcg->rkcg_assignment.all->cnt == 0) {
+                rd_kafka_topic_partition_list_destroy(
+                        rkcg->rkcg_assignment.all);
+                rkcg->rkcg_assignment.all = NULL;
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:813]: assume this is being gotten rid of

File: src/rdkafka_assignment.c [hunk 25]
```diff
@@ -0,0 +1,939 @@
...
+        }
+
+        rd_kafka_wrlock(rkcg->rkcg_rk);
+        rkcg->rkcg_c.assignment_size = rkcg->rkcg_assignment.all ?
+                rkcg->rkcg_assignment.all->cnt : 0;
+        rd_kafka_wrunlock(rkcg->rkcg_rk);
+
+        /* Any change to the assignment marks the current assignment
+         * as not lost.
+         * FIXME: Why is this? */
```
Comment (mhowlett) [inline src/rdkafka_assignment.c:825]: same reason as before

File: src/rdkafka_assignment.c (hunk 19 above)
Comment (edenhill) [reply src/rdkafka_assignment.c:568]: all of it!

File: src/rdkafka_cgrp.c [hunk 26]
```diff
@@ -148,19 +137,43 @@ rd_kafka_cgrp_incremental_unassign (rd_kafka_cgrp_t *rkcg,
...
+         RD_KAFKA_CGRP_JOIN_STATE_WAIT_REVOKE_REBALANCE_CB ||           \
+         (rkcg)->rkcg_rebalance_incr_assignment != NULL ||              \
          (rkcg)->rkcg_rebalance_rejoin)
 
 
+/**
+ * @returns true if consumer has an assignment and there is no outstanding
+ *          rebalance callback.
+ */
+#define RD_KAFKA_CGRP_MAY_REVOKE(rkcg)                  \
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:172]: this macro made it harder for me to understand what's going on than help.

File: src/rdkafka_cgrp.c [hunk 27]
```diff
@@ -148,19 +137,43 @@ rd_kafka_cgrp_incremental_unassign (rd_kafka_cgrp_t *rkcg,
...
+ *    assignment changes in its rebalance callback and then call assign().
+ * 5. An incremental rebalancing is in progress.
+ * 6. A rebalance-induced rejoin is in progress.
  */
-#define RD_KAFKA_CGRP_REBALANCING(rkcg)                                                           \
-        ((rkcg)->rkcg_join_state == RD_KAFKA_CGRP_JOIN_STATE_WAIT_JOIN  ||                        \
-         (rkcg)->rkcg_join_state == RD_KAFKA_CGRP_JOIN_STATE_WAIT_METADATA  ||                    \
-         (rkcg)->rkcg_join_state == RD_KAFKA_CGRP_JOIN_STATE_WAIT_SYNC  ||                        \
-         (rkcg)->rkcg_join_state == RD_KAFKA_CGRP_JOIN_STATE_WAIT_UNASSIGN_TO_COMPLETE  ||        \
-         (rkcg)->rkcg_join_state == RD_KAFKA_CGRP_JOIN_STATE_WAIT_INCR_UNASSIGN_TO_COMPLETE  ||   \
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:157]: did you mean to remove `RD_KAFKA_CGRP_JOIN_STATE_WAIT_INCR_UNASSIGN_TO_COMPLETE`?

File: src/rdkafka_cgrp.c [hunk 28]
```diff
@@ -797,53 +808,87 @@ rd_kafka_rebalance_op_incr (rd_kafka_cgrp_t *rkcg,
...
-
-        rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "ASSIGN",
-                     "Group \"%s\": delegating incremental %s of %d "
-                     "partition(s) to application on queue %s: %s",
-                     rkcg->rkcg_group_id->str,
-                     err == RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS ?
-                     "revoke" : "assign", partitions->cnt,
-                     rd_kafka_q_dest_name(rkcg->rkcg_q), reason);
-
         rd_kafka_cgrp_set_join_state(
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:811]: i guess minor point, but i think it was better this not going into a _CB state at all if we're ultimately going to get out of it and do this internally.

File: src/rdkafka_cgrp.c [hunk 29]
```diff
@@ -797,53 +808,87 @@ rd_kafka_rebalance_op_incr (rd_kafka_cgrp_t *rkcg,
...
+                             "unassigning all partitions and rejoining",
+                             rkcg->rkcg_group_id->str,
+                             err == RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS ?
+                             "unassign" : "assign",
+                             partitions->cnt,
+                             rd_kafka_error_string(error));
+                rd_kafka_error_destroy(error);
+
+                /* FIXME: Math; do you think this error handling is appropriate?
+                 * This will at least clean up the current mess. */
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:879]: I don't think you want to set lost here, this is just a flag for the user to check in the rb callback so they know whether the partitions in the revoke are gone due to normal CG operation or something bad. but you're clearing the partitions right?

File: src/rdkafka_cgrp.c [hunk 30]
```diff
@@ -3472,156 +3167,119 @@ static void rd_kafka_cgrp_unassign_done (rd_kafka_cgrp_t *rkcg,
...
+                     rkcg->rkcg_group_id->str,
+                     rkcg->rkcg_assignment.all ?
+                     rkcg->rkcg_assignment.all->cnt : 0);
 
-        rd_kafka_topic_partition_list_destroy(old_assignment);
+        rd_kafka_assignment_clear(rkcg);
 
-        rd_kafka_cgrp_check_unassign_done(rkcg, "unassign");
+        if (rkcg->rkcg_join_state ==
+            RD_KAFKA_CGRP_JOIN_STATE_WAIT_REVOKE_REBALANCE_CB)
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:3277]: i think this is always true?

File: src/rdkafka_cgrp.c (hunk 28 above)
Comment (mhowlett) [reply src/rdkafka_cgrp.c:811]: actually it might be logically ok - my issue might be with the name RD_KAFKA_CGRP_JOIN_STATE_WAIT_ASSIGN_REBALANCE_CB having "wait_rebalance_cb" in it.

File: src/rdkafka_cgrp.c [hunk 31]
```diff
@@ -873,71 +916,140 @@ rd_kafka_rebalance_op (rd_kafka_cgrp_t *rkcg,
...
-                        if (error)
-                                /* if set, is always fatal */
-                                rd_kafka_error_destroy(error);
-                } else
-			rd_kafka_cgrp_unassign(rkcg);
-		return rd_false;
-	}
+        rd_assert(assignment != NULL);
+
+        rd_kafka_cgrp_set_join_state(
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:935]: same thoughts as before, i in incr case, i think just leave it.

File: src/rdkafka_cgrp.c [hunk 32]
```diff
@@ -873,71 +916,140 @@ rd_kafka_rebalance_op (rd_kafka_cgrp_t *rkcg,
...
+                             "unassigning all partitions and rejoining",
+                             rkcg->rkcg_group_id->str,
+                             err == RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS ?
+                             "unassign" : "assign",
+                             rkcg->rkcg_assignment.all ?
+                             rkcg->rkcg_assignment.all->cnt : 0,
+                             rd_kafka_error_string(error));
+                rd_kafka_error_destroy(error);
+
+                /* FIXME: Math; do you think this error handling is appropriate?
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:995]: i think you don't want to set lost, like in the incr case.

File: src/rdkafka_cgrp.c [hunk 33]
```diff
@@ -873,71 +916,140 @@ rd_kafka_rebalance_op (rd_kafka_cgrp_t *rkcg,
...
+        rd_kafka_assignment_serve(rkcg);
 }
 
 
 /**
  * @brief Rejoin the group.
  */
-static void rd_kafka_cgrp_rejoin (rd_kafka_cgrp_t *rkcg) {
+static void rd_kafka_cgrp_rejoin (rd_kafka_cgrp_t *rkcg, const char *fmt, ...)
+        RD_FORMAT(printf, 2, 3);
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:1016]: i'm not familiar with what this is doing, i assume you know what you are doing

File: src/rdkafka_cgrp.c [hunk 34]
```diff
@@ -3083,135 +2997,35 @@ static rd_kafka_error_t *
...
-                        !rkcg->rkcg_assignment ? 0
-                        : rkcg->rkcg_assignment->cnt,
-                        rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
-                if (rkcg->rkcg_join_state != RD_KAFKA_CGRP_JOIN_STATE_STARTED)
-                        rd_kafka_cgrp_set_join_state(rkcg,
-                                RD_KAFKA_CGRP_JOIN_STATE_ASSIGNED);
-                goto check_rejoin;
-        }
+        if (rkcg->rkcg_join_state ==
+            RD_KAFKA_CGRP_JOIN_STATE_WAIT_ASSIGN_REBALANCE_CB) {
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:3007]: i think it should always be this

File: src/rdkafka_cgrp.c [hunk 35]
```diff
@@ -3397,7 +3086,8 @@ static void rd_kafka_cgrp_incr_unassign_done (rd_kafka_cgrp_t *rkcg,
         }
 
         /* Whether or not it was before, current assignment is now not lost. */
-        rd_atomic32_set(&rkcg->rkcg_assignment_lost, rd_false);
+        // FIXME: Should this be here or in assignment.c ?
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:3089]: afaict, seems to make sense here.

Comment (mhowlett) [reply src/rdkafka_cgrp.c:3089]: it should be here - the CG is setting assignment_t state.

also, assignment.c really shouldn't have any references to cgrp in it...

File: src/rdkafka_cgrp.c [hunk 36]
```diff
@@ -3472,156 +3167,119 @@ static void rd_kafka_cgrp_unassign_done (rd_kafka_cgrp_t *rkcg,
...
-		rd_kafka_cgrp_rejoin(rkcg);
-	}
+                rd_kafka_cgrp_rejoin(rkcg, "Unassignment done");
+        }
 
         /* Whether or not it was before, current assignment is now not lost. */
-        rd_atomic32_set(&rkcg->rkcg_assignment_lost, rd_false);
-
-	rd_kafka_cgrp_try_terminate(rkcg);
+        // FIXME: Should this be here or in assignment.c ?
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:3195]: structurally, it should be in this file

File: src/rdkafka_cgrp.c [hunk 37]
```diff
@@ -3472,156 +3167,119 @@ static void rd_kafka_cgrp_unassign_done (rd_kafka_cgrp_t *rkcg,
...
+                     rkcg->rkcg_group_id->str,
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state],
+                     RD_STR_ToF(rkcg->rkcg_rebalance_rejoin));
 
-                return;
-        }
+        switch (rkcg->rkcg_join_state)
+        {
+        case RD_KAFKA_CGRP_JOIN_STATE_WAIT_UNASSIGN_TO_COMPLETE:
+                rd_kafka_cgrp_unassign_done(rkcg, "FIXME");
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:3221]: reminder to FIXME

File: src/rdkafka_cgrp.c [hunk 38]
```diff
@@ -3472,156 +3167,119 @@ static void rd_kafka_cgrp_unassign_done (rd_kafka_cgrp_t *rkcg,
...
+                break;
 
-        if (rkcg->rkcg_join_state ==
-            RD_KAFKA_CGRP_JOIN_STATE_WAIT_INCR_UNASSIGN_TO_COMPLETE)
-                rd_kafka_cgrp_incr_unassign_done(rkcg, reason);
-        else
-	        rd_kafka_cgrp_unassign_done(rkcg, reason);
-}
+        case RD_KAFKA_CGRP_JOIN_STATE_WAIT_INCR_UNASSIGN_TO_COMPLETE:
+                rd_kafka_cgrp_incr_unassign_done(rkcg, "FIXME");
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:3225]: reminder to FIXME

File: src/rdkafka_cgrp.c [hunk 39]
```diff
@@ -5512,6 +5204,11 @@ void rd_kafka_cgrp_handle_SyncGroup (rd_kafka_cgrp_t *rkcg,
                 rd_kafka_consumer_group_metadata_destroy(cgmd);
         }
 
+        // FIXME: Remove when we're done debugging.
```
Comment (mhowlett) [inline src/rdkafka_cgrp.c:5207]: are we done debugging yet?

File: src/rdkafka_assignment.c (hunk 17 above)
Comment (mhowlett) [reply src/rdkafka_assignment.c:548]: it's maybe the case that there should be a method on assignment_t `is_stable()` or something (I don't know a good name), and `assignment_in_progress` is a cg operation.

i think the assignment abstraction you have is good.

File: src-cpp/rdkafkacpp.h (hunk 1 above)
Comment (edenhill) [reply src-cpp/rdkafkacpp.h:2686]: I agree, but in this case the amount of boilerplate we need in 4 clients is not in par with how infrequently this will be used.

File: src/rd.h (hunk 2 above)
Comment (edenhill) [reply src/rd.h:104]: If the user sees this we, as developers, have failed anyway.

File: src/rdkafka_cgrp.c (hunk 27 above)
Comment (edenhill) [reply src/rdkafka_cgrp.c:157]: Good catch!

File: src/rdkafka_cgrp.c (hunk 28 above)
Comment (edenhill) [reply src/rdkafka_cgrp.c:811]: Yeah, should probably be called WAIT_ASSIGN_CALL

File: src/rdkafka_cgrp.c (hunk 30 above)
Comment (edenhill) [reply src/rdkafka_cgrp.c:3277]: The user can call unassign at any time, regardless of state.

File: src/rdkafka_cgrp.c (hunk 33 above)
Comment (edenhill) [reply src/rdkafka_cgrp.c:1016]: Compile-time printf format checking

File: src/rdkafka_cgrp.c (hunk 34 above)
Comment (edenhill) [reply src/rdkafka_cgrp.c:3007]: User can call *assign() at any time, regardless of state.

**Comment by edenhill** [comment]
Addressed all comments
